### PR TITLE
Improve performance of leak tracker.

### PR DESCRIPTION
**Pull request by polina-c**
Move analysis of creation stack trace from start of tracking to leak filtering, to perform just for leaking objects.

### CODE REVIEW

File: pkgs/leak_tracker/lib/src/leak_tracking/_leak_filter.dart [hunk 1]
```diff
@@ -10,6 +10,12 @@ import 'primitives/model.dart';
 class LeakFilter {
   final Map<PhaseSettings, _PhaseLeakFilter> _phases = {};
 
+  /// Returns true if the leak should be reported.
+  ///
+  /// If result is true, removes [ContextKeys.startCallstack]
+  /// from [ObjectRecord.context] the call stack
```
Comment (kenzieschmoll) [inline pkgs/leak_tracker/lib/src/leak_tracking/_leak_filter.dart:16]: is this accurate? I don't see in this method where anything is "removed"

File: pkgs/leak_tracker/lib/src/leak_tracking/primitives/_test_helper_detector.dart [hunk 2]
```diff
@@ -27,13 +29,7 @@ final _startFrame = RegExp(
...
 );
 
-/// Returns whether the leak reported by [objectCreationTrace]
-/// was created by a test helper.
-///
-/// Frames, that match [exceptions] will be ignored.
-///
-/// See details on what means to be created by a test helper
-/// in doc for `LeakTesting.createdByTestHelpers`.
+@visibleForTesting
```
Comment (kenzieschmoll) [inline pkgs/leak_tracker/lib/src/leak_tracking/primitives/_test_helper_detector.dart:32]: nit: why remove dart-doc? it still helps to explain what this method does even though it is now private but visibileForTesting

File: pkgs/leak_tracker/lib/src/leak_tracking/primitives/_test_helper_detector.dart [hunk 3]
```diff
@@ -53,3 +49,37 @@ bool isCreatedByTestHelper(
...
+    );
+    _creationStack = null;
+    _exceptions = null;
+    return result;
+  }();
+
+  /// Creates instance of [CreationChecker].
+  ///
+  /// Frames, that match [exceptions] will be ignored.
+  CreationChecker(
```
Comment (kenzieschmoll) [inline pkgs/leak_tracker/lib/src/leak_tracking/primitives/_test_helper_detector.dart:81]: style nit: place constructor before all other members in class

File: pkgs/leak_tracker/lib/src/leak_tracking/primitives/_test_helper_detector.dart [hunk 4]
```diff
@@ -53,3 +49,37 @@ bool isCreatedByTestHelper(
...
+  ///
+  /// This value is cached. The first calculation of the value
+  /// is performance heavy.
+  late final bool createdByTestHelpers = () {
+    final result = isCreatedByTestHelper(
+      _creationStack!.toString(),
+      _exceptions!,
+    );
+    _creationStack = null;
+    _exceptions = null;
```
Comment (kenzieschmoll) [inline pkgs/leak_tracker/lib/src/leak_tracking/primitives/_test_helper_detector.dart:74]: why null these out? Perhaps add a comment

File: pkgs/leak_tracker/lib/src/leak_tracking/primitives/_test_helper_detector.dart [hunk 5]
```diff
@@ -53,3 +49,37 @@ bool isCreatedByTestHelper(
...
+    return result;
+  }();
+
+  /// Creates instance of [CreationChecker].
+  ///
+  /// Frames, that match [exceptions] will be ignored.
```
Comment (kenzieschmoll) [inline pkgs/leak_tracker/lib/src/leak_tracking/primitives/_test_helper_detector.dart:80]: nit: to make this more clear "Stack frames in [creationStack] that match any of [exceptions] will be ignored."

File: pkgs/leak_tracker/lib/src/leak_tracking/_leak_filter.dart [hunk 6]
```diff
@@ -28,24 +34,35 @@ class _PhaseLeakFilter {
...
     }
+
+    // Check for test helpers should happen only in case of leak because
+    // it is performance heavy.
+    final shouldCheckCreator =
+        phase.ignoredLeaks.createdByTestHelpers && result;
+
+    if (!shouldCheckCreator) return result;
+
+    return !(record.creationChecker?.createdByTestHelpers ?? false);
```
Comment (kenzieschmoll) [inline pkgs/leak_tracker/lib/src/leak_tracking/_leak_filter.dart:61]: nit: to make this more readable, consider:
```
final createdByTestHelpers = record.creationChecker?.createdByTestHelpers ?? false;
return !createdByTestHelpers;
```

File: pkgs/leak_tracker/lib/src/leak_tracking/_leak_filter.dart [hunk 7]
```diff
@@ -28,24 +34,35 @@ class _PhaseLeakFilter {
...
   }
 
-  bool _shouldReport(
+  /// Returns true if the leak should not be ignored because of type or class.
```
Comment (kenzieschmoll) [inline pkgs/leak_tracker/lib/src/leak_tracking/_leak_filter.dart:64]: nit: remove the double negative "not be ignored". 

"Returns whether the leak should be reported based on its type and class."

File: pkgs/leak_tracker/lib/src/leak_tracking/_object_record.dart [hunk 8]
```diff
@@ -26,6 +28,8 @@ class ObjectRecord {
   /// the object is already GCed and thus there is no access to its code.
   final IdentityHashCode code;
 
+  final CreationChecker? creationChecker;
```
Comment (kenzieschmoll) [inline pkgs/leak_tracker/lib/src/leak_tracking/_object_record.dart:31]: dart doc. Prefer to add dart doc for all of these public members while you are here.

**Review by kenzieschmoll — COMMENTED** [review]
consider adding a test to ensure that the test helper check does not run from `shouldReport` when this is false:
```
final shouldCheckCreator =
        phase.ignoredLeaks.createdByTestHelpers && result;
```

**Comment by polina-c** [comment]
> consider adding a test to ensure that the test helper check does not run from `shouldReport` when this is false:
> 
> ```
> final shouldCheckCreator =
>         phase.ignoredLeaks.createdByTestHelpers && result;
> ```

Added  TODO

File: pkgs/leak_tracker/lib/src/leak_tracking/_object_record.dart [hunk 9]
```diff
@@ -28,6 +29,10 @@ class ObjectRecord {
   /// the object is already GCed and thus there is no access to its code.
   final IdentityHashCode code;
 
+  /// [CreationChecker] that contains knowledge about creation.
+  ///
+  /// Is not used in the record, but can be used
+  /// that can be used by owners of this object.
```
Comment (kenzieschmoll) [inline pkgs/leak_tracker/lib/src/leak_tracking/_object_record.dart:35]: remove duplicate "that can be used"

**Review by kenzieschmoll — APPROVED** [review]

**Pull request merged**